Fix item picker suggests to select items user does not have write access to

kulyk opened the pull request:
Fixes #15281 

### Problem

There is a number of flows when a user adds-creates something somewhere and they need to specify the "target location". For example:

* select a dashboard to save question to
* select a collection to save dashboard to

Users might have different permissions to these resources. Right now, when we display e.g. a list of collections, we just show the whole list. So it's possible to select a collection you won't be able to actually modify and you'll end up with an error somewhere on UI.

This PR ensures we filter this things according to users' permissions

Also fixes a bug when selecting a collection mentioned [here](https://github.com/metabase/metabase/issues/15281#issuecomment-804112093)

kulyk committed:
Test adding question to dashboard

kulyk committed:
Test collections filtering when adding a question

When adding a question to dashboard,
we need to display collections a user has "write" access to.
Collection with "read" access have to be hidden

kulyk committed:
Fix adding question to dashboard without access

kulyk committed:
Add a note about permissions test suite

kulyk committed:
Move question permission tests to collection suite

kulyk committed:
Revert initial collections filtering

kulyk committed:
Fix adding question to dashboard without access

kulyk committed:
Remove redundant state field

kulyk committed:
Enable #15281 issue repro test

kulyk committed:
Remove requireCollectionWritePermission prop

kulyk committed:
Filter items user doesn't have `write` access to

kulyk committed:
Fix permission tests

kulyk committed:
Fix dashboard test

nemanjaglumac commented on frontend/test/metabase/scenarios/collections/permissions.cy.spec.js:
```diff
@@ -1,3 +1,15 @@
+/**
+ * FYI, this test suite contains permission tests for different pages
+ *
+ * - Collections (/collection/root)
+ * - Dashboard (/dashboard/:id)
+ * - Question (/question/:id)
+ *
+ * It's a WIP and most likely it will be split later,
+ * when we're sure about our testing strategy for permissions
+ * See discussion: https://github.com/metabase/metabase/pull/15573
+ */
+
```
Awesome, thanks for this comment. It's super helpful for the future reference.

nemanjaglumac commented on frontend/test/metabase/scenarios/collections/permissions.cy.spec.js:
```diff
@@ -285,6 +297,43 @@ describe("collection permissions", () => {
+
+                describe("adding question to dashboard", () => {
+                  beforeEach(() => {
+                    cy.skipOn(user === "nodata");
```
Why do we skip every test for "nodata" user?

kulyk replied:
Ouch, haven't noticed it. Removing, thanks!

kulyk committed:
Fix part of permission tests disabled for nodata user

kulyk committed:
Bring back issue reference to Cypress test

flamber reviewed (commented):
As for #15280:

This seems to remove the `...` picker menu for collection view-user, so they cannot duplicate a dashboard (and question in the future, 15255).
Also removed the option, when going to a dashboard as view-user. For questions, this option hasn't been visible yet, but it should.

A view-user should be able to duplicate items from a view-collection and save it to a collection they have curate-access to.
A view-user should not see options that can change the item in a view-collection like "Archive" or "Change title and description" (or "Move" etc).

flamber commented:
As for #15281:

Going to a collection that is revoked or view-only, and click "New dashboard", then it still offers to save to that location by default.
![image](https://user-images.githubusercontent.com/1447303/114869552-690ee980-9df7-11eb-90ca-fe397f8f1806.png)

flamber commented:
But it seems to fix #8996

paulrosenzweig commented on frontend/src/metabase/containers/ItemPicker.jsx:
```diff
@@ -76,6 +76,32 @@ export default class ItemPicker extends React.Component {
     }
   }
 
+  _checkHasWritePermissionForItem(item, models) {
```
We don't use this naming convention with leading underscores much.

paulrosenzweig commented on frontend/src/metabase/containers/ItemPicker.jsx:
```diff
@@ -76,6 +76,32 @@ export default class ItemPicker extends React.Component {
+  }
+
+  _checkCollectionHasChildWithWritePermission(collection) {
+    const hasChildren = collection.children.length > 0;
+    if (!hasChildren) {
+      return collection.can_write;
+    }
+    return collection.children.some(child =>
+      this._checkCollectionHasChildWithWritePermission(child),
+    );
```
What about collections that are themselves writable, have children, but none of the children are writable? It seems like the logic should be `this is writable || this has writable children`. I haven't tested it though so I might be missing something.

kulyk replied:
You're right, they get hidden in this case. Good catch, thank you!

paulrosenzweig reviewed (approved):
Code looks good! I left one question about behavior, but I didn't spot anything besides that.

paulrosenzweig commented:
> A view-user should be able to duplicate items from a view-collection and save it to a collection they have curate-access to.

I think that change was done [here](https://github.com/metabase/metabase/pull/15573/files#diff-39f914fbd0b225abdcaa6f8848dadeb49298c1efec4cc759f8f9ee91f2c91a2cR310).

kulyk commented:
> Going to a collection that is revoked or view-only, and click "New dashboard", then it still offers to save to that location by default

@flamber thanks for catching that 💪 Will fix

kulyk commented:
> > A view-user should be able to duplicate items from a view-collection and save it to a collection they have curate-access to.
> 
> I think that change was done [here](https://github.com/metabase/metabase/pull/15573/files#diff-39f914fbd0b225abdcaa6f8848dadeb49298c1efec4cc759f8f9ee91f2c91a2cR310).

Yes, you're right, I think it should be addressed in another PR then

kulyk committed:
Remove underscore prefixes for component methods

kulyk committed:
Test offers saving dashboard to opened collection

kulyk committed:
Fix tests nested incorrectly

kulyk committed:
Split dashboard permission test

kulyk committed:
Fix suggest saving items to read-only collections

kulyk committed:
Fix collection permission filtering

See comment:
https://github.com/metabase/metabase/pull/15613#discussion_r614460504

kulyk committed:
Move comment

kulyk committed:
Fix test failing due to fixed collection suggestion

nemanjaglumac commented on frontend/test/metabase/scenarios/collections/permissions.cy.spec.js:
```diff
@@ -24,6 +36,21 @@ describe("collection permissions", () => {
+                  });
+                  cy.icon("add").click();
+                  cy.findByText("New dashboard").click();
+                  cy.get(".Modal")
+                    .findByText("Second collection")
+                    .should("exist");
```
```suggestion
                  cy.get(".AdminSelect").findByText("Second collection");
```
No need to use explicit `.should("exist")` assertion. `cy.findByText()` fails if it doesn't find a string and/or if it finds multiple strings so implicit assertion is always preferred.

Btw, I suggested we check for `.AdminSelect` because it is more specific and it makes it more clear to anyone who reads this test later what was the intention behind this assertion.

kulyk replied:
Thanks for your feedback! Applied the change in my own commit not to trigger the CI once again

nemanjaglumac commented on frontend/test/metabase/scenarios/collections/permissions.cy.spec.js:
```diff
@@ -285,6 +312,42 @@ describe("collection permissions", () => {
                   cy.location("pathname").should("eq", "/collection/root");
                   cy.findByText("Orders").should("not.exist");
                 });
+
+                describe("adding question to dashboard", () => {
+                  beforeEach(() => {
+                    popover()
+                      .findByText("Add to dashboard")
+                      .click();
+                  });
+
+                  it("should list collections user has `write` access to", () => {
+                    cy.get(".Modal")
+                      .findByText("Orders in a dashboard")
+                      .should("exist");
+                  });
+
+                  it("should list dashboards user has `write` access to when using search", () => {
+                    cy.get(".Modal").within(() => {
+                      cy.icon("search").click();
+                      cy.findByPlaceholderText("Search").type(
+                        "Orders in{Enter}",
+                      );
+                      cy.findByText("Orders in a dashboard").should("exist");
+                    });
+                  });
+
+                  it("should be able to add question to dashboard", () => {
+                    cy.get(".Modal")
+                      .as("modal")
+                      .findByText("Orders in a dashboard")
+                      .click();
+
+                    cy.get("@modal").should("not.exist");
+                    // By default, the dashboard contains one question
+                    // After we add a new one, we check there are two questions now
+                    cy.get(".DashCard").should("have.length", 2);
+                  });
+                });
```
All of this can and should be a single test without an additional describe block. Again, the same suggestion applies as in the comment before - no need to explicitly use`.should("exist")`.

If you take a look at the test runner sidebar and the generated titles, you'll see that we don't have to explicitly say "user has `write` access to". Every member from `curate` group has write access to all collections.
![image](https://user-images.githubusercontent.com/31325167/115278034-422d1c00-a145-11eb-8d1e-a2d3adc7d4d7.png)


Also, I don't see any benefit in using/testing the search in this particular test because "Orders in a dashboard" was already shown in the modal. Testing search makes more sense for users who might not have write access to some collection because we want to make sure they are not even offered to them.

kulyk replied:
Awesome! Merged the tests, removed explicit `"exists"` assertion, removed the "has write access" notes

kulyk committed:
Remove `should("exist")` from Cypress tests

kulyk committed:
Merge "adding question to dashboard" tests

kulyk committed:
Merge similar permission tests

kulyk committed:
Merge similar tests into one

kulyk committed:
Use sidebar test ID in permissions test

kulyk committed:
Select by .AdminSelect

kulyk committed:
Remove redundant search test case

kulyk commented:
@nemanjaglumac thanks for your feedback, I guess everything is addressed and fixed now. Please take a look again

nemanjaglumac commented on frontend/test/metabase/scenarios/question/native.cy.spec.js:
```diff
@@ -229,7 +229,6 @@ describe("scenarios > question > native", () => {
     cy.findByText("Simple question").click();
     popover().within(() => {
       cy.findByText("Saved Questions").click();
-      cy.findByText("Robert Tableton's Personal Collection").click();
```
Everything looks good. Why was this change needed?

kulyk replied:
That's great you asked, I was not 100% sure and went to double-check and there is actually an unexpected behavior here

nemanjaglumac reviewed (approved):
Left a comment for `native.cy.spec.js` but overall looks pretty good. Thank you for addressing all the comments! 🙏

kulyk committed:
Revert native query test

kulyk committed:
Add React list key to ItemPicker items

kulyk committed:
Add collection suggestions tests

kulyk committed:
Configure Form's `overwriteOnInitialValuesChange`

kulyk committed:
Wrap CollectionSelect with `@Collection.loadList`

When suggesting an initial collection,
we need to check a user has `write` access to it.
For that, collection objects have to be present in Redux store,
so we can retrieve a collection by ID and check the `can_write` flag

kulyk committed:
Allow modifying CreateDashboardModal's onSave prop

kulyk committed:
Fix collection suggestsions for new dashboard

kulyk committed:
Fix collection suggestions when copying dashboards

kulyk committed:
Add defaultProps to Form

kulyk committed:
Fix SaveQuestionModal unit test

kulyk committed:
Fix SaveQuestionModal collection suggestion

kulyk commented:
@paulrosenzweig @nemanjaglumac @flamber, thanks for the feedback you already shared. I found and fixed a few edge-cases here. Can you please give this PR one more look?

A bit of context on the recent changes:

After some fixes, [one of the native query tests failed](https://app.circleci.com/pipelines/github/metabase/metabase/15162/workflows/cafe15e3-6733-4479-a9a8-baa08e0c1942/jobs/580745/tests#failed-test-0). Here is the scenario of the failed test:

* a normal user creates a native query they will use later via the "Saved Questions" menu
* the user saves the query with a dialog changed in this PR
* the dialog has to suggest to save the question to the user's personal collection, but it suggests "Our analytics."

<img width="1437" src="https://user-images.githubusercontent.com/17258145/116224280-27683200-a759-11eb-8633-2dd29229cc80.png">

⬆️ This is a screenshot of the correct behavior. In the failing scenario, the popup suggests saving the question to "Our analytics".

The regression happened because of a change in [`getInitialCollectionId` selector](https://github.com/metabase/metabase/blob/e53c482a35dca1894413cbefa99f27cfafbee9fa/frontend/src/metabase/entities/collections.js#L91). When a user has to select a dashboard somewhere, the initial value is usually received from this selector. It collects a list of collection IDs in prioritized order (e.g., the top priority is the collection ID passed to the component via props, then via router params, etc. If nothing is passed, the selector fallbacks to the user's collection). Once the list is ready, we try to return the highest-priority option (closest to the top of the list).

This PR changes the selector, so we now check if a user has to write access to the collection to avoid irrelevant suggestions. That means, when the selector is called, **collections have to be present in the Redux store** to perform the check.

The test failed because it covers the behavior when `getInitialCollectionId` is called before the collection list is retrieved and saved to Redux. In this particular scenario, `getInitialCollectionId` would return user’s personal collection ID. But the selector will now take the collection object by ID to check its `can_write` flag. As the collections dictionary is empty, we can't check the `can_write,` and the selector will return `undefined`, making the modal fallback to the root collection ("Our analytics")

Ideally, we should fix the back-end API so the front end can retrieve only collections with write access. But for now, I wrapped collection selection input with `Collection.loadList` to guarantee each time user is offered to select a collection, and Redux store contains a list of collections. Also, I had to fix a few forms here and there, so they refresh when `initialCollectionId` is changed.

Please take a look and LMK if there is an easier way to achieve the same behavior 🙏

nemanjaglumac commented on frontend/test/metabase/scenarios/collections/permissions.cy.spec.js:
```diff
@@ -367,22 +408,24 @@ describe("collection permissions", () => {
               popover()
                 .findByText("Duplicate this item")
-                .should("exist");
+                .click();
+              cy.get(".AdminSelect").findByText(
+                `${first_name} ${last_name}'s Personal Collection`,
+              );
```
Nice additional check!  🙏

nemanjaglumac commented on frontend/test/metabase/scenarios/collections/permissions.cy.spec.js:
```diff
@@ -401,6 +444,45 @@ describe("collection permissions", () => {
+                });
+                cy.url().should("match", /\/dashboard\/\d+$/);
+                saveDashboard();
+                cy.get(".Entity").findByText(personalCollection);
```
Nitpick: `. DashboardHeader` conveys more meaning. I had no idea whatsoever what is `.Entity` so I had to build the test and observe what's going on in the Cypress runner.

Don't have to change it in this PR. Just a general note.

kulyk replied:
Fixed, it's much better now. Thank you 🙏

nemanjaglumac commented on frontend/test/metabase/scenarios/collections/permissions.cy.spec.js:
```diff
@@ -401,6 +444,45 @@ describe("collection permissions", () => {
             });
 
+            describe("managing question from the question's edit dropdown", () => {
+              it("should not be offered to save question in collections they have `read` access to", () => {
```
This is not "saving a question in a collection". We're adding a question to the dashboard that lives in a collection we have 'read' access to. The difference between the two scenarios is huge.

kulyk replied:
My bad. Found and fixed one more inconsistent test name. Thank you!

nemanjaglumac reviewed (approved):
Left a few comments, but tests overall look good.
Thank you for working hard on this! 

Didn't review the frontend, so best wait for other reviews as well.

paulrosenzweig reviewed (approved):
Left some comments, but it looks good!

One thought: Looking at this again, I'm somewhat nervous about unanticipated bugs. It touches a lot of important modals. Given that it's not a super urgent bug to fix, it might be worth shipping this with 40 so we have more time to catch anything.

kulyk commented on frontend/src/metabase/collections/components/CollectionCopyEntityModal.jsx:
```diff
@@ -9,18 +10,32 @@ import { entityTypeForObject } from "metabase/schema";
       entityType={entityTypeForObject(entityObject)}
-      entityObject={entityObject}
+      entityObject={{
+        ...entityObject,
+        collection_id: initialCollectionId,
```
Absolutely! Thanks, missed that

kulyk committed:
Simplify CollectionSelect wrapper

kulyk committed:
Fix dashboard header test selector

kulyk committed:
Rename permission tests

kulyk committed:
Mock HTTP requests at SaveQuestionModal test

kulyk committed:
Pass correct params to initialCollectionId

kulyk commented:
@paulrosenzweig 

> Looking at this again, I'm somewhat nervous about unanticipated bugs. It touches a lot of important modals. Given that it's not a super urgent bug to fix, it might be worth shipping this with 40 so we have more time to catch anything.

Agree here. I'll rebase it on master them and share a heads up as the PR is merged
